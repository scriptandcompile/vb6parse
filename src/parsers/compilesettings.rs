use num_enum::TryFromPrimitive;
use serde::Serialize;

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum UnroundedFloatingPoint {
    #[default]
    DoNotAllow = 0,
    Allow = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum PentiumFDivBugCheck {
    CheckPentiumFDivBug = 0,
    #[default]
    NoPentiumFDivBugCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum BoundsCheck {
    #[default]
    CheckBounds = 0,
    NoBoundsCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum OverflowCheck {
    #[default]
    CheckOverflow = 0,
    NoOverflowCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum FloatingPointErrorCheck {
    #[default]
    CheckFloatingPointError = 0,
    NoFloatingPointErrorCheck = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum CodeViewDebugInfo {
    #[default]
    NotCreated = 0,
    Created = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum FavorPentiumPro {
    #[default]
    False = 0,
    True = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum Aliasing {
    #[default]
    AssumeAliasing = 0,
    AssumeNoAliasing = -1,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Default, TryFromPrimitive)]
#[repr(i16)]
pub enum OptimizationType {
    #[default]
    FavorFastCode = 0,
    FavorSmallCode = 1,
    NoOptimization = 2,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub enum CompilationType {
    // 0
    NativeCode {
        optimization_type: OptimizationType,
        favor_pentium_pro: FavorPentiumPro,
        code_view_debug_info: CodeViewDebugInfo,
        aliasing: Aliasing,
        bounds_check: BoundsCheck,
        overflow_check: OverflowCheck,
        floating_point_check: FloatingPointErrorCheck,
        pentium_fdiv_bug_check: PentiumFDivBugCheck,
        unrounded_floating_point: UnroundedFloatingPoint,
    },
    // -1, default
    PCode,
}
