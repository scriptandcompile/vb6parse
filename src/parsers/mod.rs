//! This module contains the file parsers for the Visual Basic 6 language.
//!
//! The major components of this module are:
//!
//! `VB6Stream` - The main parsing structure for VB6 files. It holds the
//! bytes for a file and provides basic methoda for iterating over the bytes as
//! well as tracking the line, column, and character position of the current byte.
//!
//! `VB6Project` - a structure for holding the parsed information from a VB6
//! project file.
//!
//! `VB6ClassFile` - a structure for holding the parsed information from a VB6
//! class file.
//!
//! `VB6FormFile` - a structure for holding the parsed information from a VB6
//! form file.
//!
//! `VB6ModuleFile` - a structure for holding the parsed information from a VB6
//! module file.
//!

mod header;
mod vb6stream;

pub mod class;
pub mod compilesettings;
pub mod form;
pub mod module;
pub mod objectreference;
pub mod project;
pub mod properties;
pub mod vb6;

pub use class::{
    FileUsage, MtsStatus, Persistance, VB6ClassFile, VB6ClassHeader, VB6ClassProperties,
    VB6ClassVersion,
};

pub use compilesettings::{
    BoundsCheck, CompilationType, FloatingPointErrorCheck, OverflowCheck, PentiumFDivBugCheck,
    UnroundedFloatingPoint,
};
pub use form::{resource_file_resolver, VB6FormFile};
pub use header::{VB6FileAttributes, VB6FileFormatVersion};
pub use module::VB6ModuleFile;

pub use properties::Properties;

pub use objectreference::VB6ObjectReference;

pub use project::{
    CompileTargetType, VB6Project, VB6ProjectClass, VB6ProjectModule, VB6ProjectReference,
};

pub use vb6::{is_english_code, vb6_parse};

pub use vb6stream::VB6Stream;
