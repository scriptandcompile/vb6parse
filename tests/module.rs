use vb6parse::parsers::VB6ModuleFile;

#[test]
fn artificial_life_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Artificial-life/Declarations.bas");

    let module_file = match VB6ModuleFile::parse("Declarations.bas".to_owned(), module_file_bytes) {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Declarations.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn game_physics_basic_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Game-physics-basic/Physics_Logic.bas");

    let module_file = match VB6ModuleFile::parse("Physics_Logic.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Physics_Logic.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn histograms_advanced_module_load() {
    let module_file_bytes =
        include_bytes!("./data/vb6-code/Histograms-advanced/mod_PublicVars.bas");

    let module_file = match VB6ModuleFile::parse("mod_PublicVars.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'mod_PublicVars.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn histograms_basic_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Histograms-basic/mod_PublicVars.bas");

    let module_file = match VB6ModuleFile::parse("mod_PublicVars.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'mod_PublicVars.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn levels_effect_module_load() {
    let module_file_bytes = include_bytes!("./data/vb6-code/Levels-effect/mod_PublicVars.bas");

    let module_file = match VB6ModuleFile::parse("mod_PublicVars.bas".to_owned(), module_file_bytes)
    {
        Ok(module_file) => module_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'mod_PublicVars.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module_file);
}

#[test]
fn map_editor_2d_module_load() {
    let module1_file_bytes = include_bytes!("./data/vb6-code/Map-editor-2D/Subs.bas");

    let module1_file = match VB6ModuleFile::parse("Subs.bas".to_owned(), module1_file_bytes) {
        Ok(module1_file) => module1_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Subs.bas' module file");
        }
    };

    let module2_file_bytes = include_bytes!("./data/vb6-code/Map-editor-2D/Declarations.bas");

    let module2_file = match VB6ModuleFile::parse("Declarations.bas".to_owned(), module2_file_bytes)
    {
        Ok(module2_file) => module2_file,
        Err(e) => {
            eprintln!("{e}");
            panic!("Failed to parse 'Declarations.bas' module file");
        }
    };

    insta::assert_yaml_snapshot!(module1_file);
    insta::assert_yaml_snapshot!(module2_file);
}
