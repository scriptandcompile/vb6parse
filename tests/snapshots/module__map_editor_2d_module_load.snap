---
source: tests/module.rs
expression: module1_file
---
name:
  - 34
  - 68
  - 101
  - 99
  - 108
  - 97
  - 114
  - 97
  - 116
  - 105
  - 111
  - 110
  - 95
  - 77
  - 111
  - 100
  - 117
  - 108
  - 101
  - 34
  - 13
  - 10
tokens:
  - Comment:
      - 39
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 67
      - 111
      - 112
      - 121
      - 114
      - 105
      - 103
      - 104
      - 116
      - 32
      - 50
      - 48
      - 49
      - 56
      - 32
      - 98
      - 121
      - 32
      - 84
      - 97
      - 110
      - 110
      - 101
      - 114
      - 32
      - 72
      - 101
      - 108
      - 108
      - 97
      - 110
      - 100
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 32
      - 119
      - 119
      - 119
      - 46
      - 116
      - 97
      - 110
      - 110
      - 101
      - 114
      - 104
      - 101
      - 108
      - 108
      - 97
      - 110
      - 100
      - 46
      - 99
      - 111
      - 109
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 68
      - 111
      - 99
      - 117
      - 109
      - 101
      - 110
      - 116
      - 97
      - 116
      - 105
      - 111
      - 110
      - 32
      - 102
      - 111
      - 114
      - 32
      - 116
      - 104
      - 105
      - 115
      - 32
      - 112
      - 114
      - 111
      - 106
      - 101
      - 99
      - 116
      - 32
      - 99
      - 97
      - 110
      - 32
      - 98
      - 101
      - 32
      - 102
      - 111
      - 117
      - 110
      - 100
      - 32
      - 97
      - 116
      - 32
      - 104
      - 116
      - 116
      - 112
      - 115
      - 58
      - 47
      - 47
      - 116
      - 97
      - 110
      - 110
      - 101
      - 114
      - 104
      - 101
      - 108
      - 108
      - 97
      - 110
      - 100
      - 46
      - 99
      - 111
      - 109
      - 47
      - 99
      - 111
      - 100
      - 101
      - 47
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 115
      - 111
      - 117
      - 114
      - 99
      - 101
      - 32
      - 99
      - 111
      - 100
      - 101
      - 32
      - 105
      - 110
      - 32
      - 116
      - 104
      - 105
      - 115
      - 32
      - 112
      - 114
      - 111
      - 106
      - 101
      - 99
      - 116
      - 32
      - 105
      - 115
      - 32
      - 108
      - 105
      - 99
      - 101
      - 110
      - 115
      - 101
      - 100
      - 32
      - 117
      - 110
      - 100
      - 101
      - 114
      - 32
      - 97
      - 32
      - 83
      - 105
      - 109
      - 112
      - 108
      - 105
      - 102
      - 105
      - 101
      - 100
      - 32
      - 66
      - 83
      - 68
      - 32
      - 108
      - 105
      - 99
      - 101
      - 110
      - 115
      - 101
      - 46
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 32
      - 70
      - 111
      - 114
      - 32
      - 109
      - 111
      - 114
      - 101
      - 32
      - 105
      - 110
      - 102
      - 111
      - 114
      - 109
      - 97
      - 116
      - 105
      - 111
      - 110
      - 44
      - 32
      - 112
      - 108
      - 101
      - 97
      - 115
      - 101
      - 32
      - 114
      - 101
      - 118
      - 105
      - 101
      - 119
      - 32
      - 76
      - 73
      - 67
      - 69
      - 78
      - 83
      - 69
      - 46
      - 109
      - 100
      - 32
      - 97
      - 116
      - 32
      - 104
      - 116
      - 116
      - 112
      - 115
      - 58
      - 47
      - 47
      - 103
      - 105
      - 116
      - 104
      - 117
      - 98
      - 46
      - 99
      - 111
      - 109
      - 47
      - 116
      - 97
      - 110
      - 110
      - 101
      - 114
      - 104
      - 101
      - 108
      - 108
      - 97
      - 110
      - 100
      - 47
      - 116
      - 104
      - 100
      - 99
      - 45
      - 99
      - 111
      - 100
      - 101
      - 47
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 73
      - 102
      - 32
      - 121
      - 111
      - 117
      - 32
      - 102
      - 105
      - 110
      - 100
      - 32
      - 116
      - 104
      - 105
      - 115
      - 32
      - 99
      - 111
      - 100
      - 101
      - 32
      - 117
      - 115
      - 101
      - 102
      - 117
      - 108
      - 44
      - 32
      - 112
      - 108
      - 101
      - 97
      - 115
      - 101
      - 32
      - 99
      - 111
      - 110
      - 115
      - 105
      - 100
      - 101
      - 114
      - 32
      - 97
      - 32
      - 115
      - 109
      - 97
      - 108
      - 108
      - 32
      - 100
      - 111
      - 110
      - 97
      - 116
      - 105
      - 111
      - 110
      - 32
      - 116
      - 111
      - 32
      - 104
      - 116
      - 116
      - 112
      - 115
      - 58
      - 47
      - 47
      - 119
      - 119
      - 119
      - 46
      - 112
      - 97
      - 121
      - 112
      - 97
      - 108
      - 46
      - 109
      - 101
      - 47
      - 84
      - 97
      - 110
      - 110
      - 101
      - 114
      - 72
      - 101
      - 108
      - 108
      - 97
      - 110
      - 100
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
      - 42
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 83
      - 116
      - 114
      - 101
      - 116
      - 99
      - 104
      - 66
      - 108
      - 116
      - 32
      - 100
      - 101
      - 99
      - 108
      - 97
      - 114
      - 97
      - 116
      - 105
      - 111
      - 110
      - 32
      - 40
      - 116
      - 104
      - 101
      - 32
      - 71
      - 68
      - 73
      - 51
      - 50
      - 32
      - 97
      - 108
      - 116
      - 101
      - 114
      - 110
      - 97
      - 116
      - 105
      - 118
      - 101
      - 32
      - 116
      - 111
      - 32
      - 86
      - 66
      - 39
      - 115
      - 32
      - 80
      - 97
      - 105
      - 110
      - 116
      - 80
      - 105
      - 99
      - 116
      - 117
      - 114
      - 101
      - 41
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - DeclareKeyword:
      - 68
      - 101
      - 99
      - 108
      - 97
      - 114
      - 101
  - Whitespace:
      - 32
  - FunctionKeyword:
      - 70
      - 117
      - 110
      - 99
      - 116
      - 105
      - 111
      - 110
  - Whitespace:
      - 32
  - Identifier:
      - 83
      - 116
      - 114
      - 101
      - 116
      - 99
      - 104
      - 66
      - 108
      - 116
  - Whitespace:
      - 32
  - LibKeyword:
      - 76
      - 105
      - 98
  - Whitespace:
      - 32
  - StringLiteral:
      - 34
      - 71
      - 68
      - 73
      - 51
      - 50
      - 34
  - Whitespace:
      - 32
  - LeftParanthesis:
      - 40
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 104
      - 68
      - 101
      - 115
      - 116
      - 68
      - 67
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 120
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 89
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 110
      - 87
      - 105
      - 100
      - 116
      - 104
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 110
      - 72
      - 101
      - 105
      - 103
      - 104
      - 116
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 104
      - 83
      - 114
      - 99
      - 68
      - 67
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 120
      - 83
      - 114
      - 99
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 121
      - 83
      - 114
      - 99
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 67
      - 108
      - 105
      - 112
      - 88
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 67
      - 108
      - 105
      - 112
      - 89
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - ByValKeyword:
      - 66
      - 121
      - 86
      - 97
      - 108
  - Whitespace:
      - 32
  - Identifier:
      - 82
      - 97
      - 115
      - 116
      - 101
      - 114
      - 79
      - 112
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - RightParanthesis:
      - 41
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 65
      - 110
      - 32
      - 97
      - 114
      - 114
      - 97
      - 121
      - 44
      - 32
      - 119
      - 104
      - 105
      - 99
      - 104
      - 32
      - 119
      - 101
      - 39
      - 108
      - 108
      - 32
      - 117
      - 115
      - 101
      - 32
      - 116
      - 111
      - 32
      - 116
      - 114
      - 97
      - 99
      - 107
      - 32
      - 111
      - 117
      - 114
      - 32
      - 109
      - 97
      - 112
      - 32
      - 116
      - 105
      - 108
      - 101
      - 32
      - 118
      - 97
      - 108
      - 117
      - 101
      - 115
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 77
      - 97
      - 112
      - 65
      - 114
      - 114
      - 97
      - 121
  - LeftParanthesis:
      - 40
  - RightParanthesis:
      - 41
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - ByteKeyword:
      - 66
      - 121
      - 116
      - 101
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 76
      - 111
      - 111
      - 112
      - 32
      - 118
      - 97
      - 114
      - 105
      - 97
      - 98
      - 108
      - 101
      - 115
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 120
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - Identifier:
      - 89
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 99
      - 117
      - 114
      - 114
      - 101
      - 110
      - 116
      - 108
      - 121
      - 32
      - 115
      - 101
      - 108
      - 101
      - 99
      - 116
      - 101
      - 100
      - 32
      - 112
      - 111
      - 115
      - 105
      - 116
      - 105
      - 111
      - 110
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 67
      - 117
      - 114
      - 88
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - Identifier:
      - 67
      - 117
      - 114
      - 89
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 118
      - 105
      - 101
      - 119
      - 112
      - 111
      - 114
      - 116
      - 32
      - 112
      - 111
      - 115
      - 105
      - 116
      - 105
      - 111
      - 110
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 86
      - 105
      - 101
      - 119
      - 88
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - Identifier:
      - 86
      - 105
      - 101
      - 119
      - 89
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 115
      - 105
      - 122
      - 101
      - 32
      - 111
      - 102
      - 32
      - 116
      - 104
      - 101
      - 32
      - 109
      - 97
      - 112
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 83
      - 105
      - 122
      - 101
      - 88
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - IntegerKeyword:
      - 73
      - 110
      - 116
      - 101
      - 103
      - 101
      - 114
  - Comma:
      - 44
  - Whitespace:
      - 32
  - Identifier:
      - 83
      - 105
      - 122
      - 101
      - 89
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - IntegerKeyword:
      - 73
      - 110
      - 116
      - 101
      - 103
      - 101
      - 114
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 87
      - 104
      - 101
      - 116
      - 104
      - 101
      - 114
      - 32
      - 97
      - 32
      - 109
      - 111
      - 117
      - 115
      - 101
      - 32
      - 98
      - 117
      - 116
      - 116
      - 111
      - 110
      - 32
      - 105
      - 115
      - 32
      - 117
      - 112
      - 32
      - 111
      - 114
      - 32
      - 100
      - 111
      - 119
      - 110
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 77
      - 111
      - 117
      - 115
      - 101
      - 83
      - 116
      - 97
      - 116
      - 101
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - BooleanKeyword:
      - 66
      - 111
      - 111
      - 108
      - 101
      - 97
      - 110
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 99
      - 117
      - 114
      - 114
      - 101
      - 110
      - 116
      - 108
      - 121
      - 32
      - 115
      - 101
      - 108
      - 101
      - 99
      - 116
      - 101
      - 100
      - 32
      - 112
      - 105
      - 99
      - 116
      - 117
      - 114
      - 101
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 67
      - 117
      - 114
      - 80
      - 105
      - 99
      - 73
      - 110
      - 100
      - 101
      - 120
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Comma:
      - 44
  - Whitespace:
      - 32
  - Identifier:
      - 67
      - 117
      - 114
      - 80
      - 105
      - 99
      - 73
      - 110
      - 100
      - 101
      - 120
      - 50
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - LongKeyword:
      - 76
      - 111
      - 110
      - 103
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 84
      - 104
      - 101
      - 32
      - 122
      - 111
      - 111
      - 109
      - 32
      - 118
      - 97
      - 108
      - 117
      - 101
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 90
      - 111
      - 111
      - 109
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - IntegerKeyword:
      - 73
      - 110
      - 116
      - 101
      - 103
      - 101
      - 114
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
  - Comment:
      - 39
      - 87
      - 104
      - 101
      - 116
      - 104
      - 101
      - 114
      - 32
      - 111
      - 114
      - 32
      - 110
      - 111
      - 116
      - 32
      - 116
      - 111
      - 32
      - 117
      - 115
      - 101
      - 32
      - 83
      - 116
      - 114
      - 101
      - 116
      - 99
      - 104
      - 66
      - 108
      - 116
      - 32
      - 105
      - 110
      - 32
      - 112
      - 108
      - 97
      - 99
      - 101
      - 32
      - 111
      - 102
      - 32
      - 80
      - 97
      - 105
      - 110
      - 116
      - 80
      - 105
      - 99
      - 116
      - 117
      - 114
      - 101
  - Newline:
      - 13
      - 10
  - PublicKeyword:
      - 80
      - 117
      - 98
      - 108
      - 105
      - 99
  - Whitespace:
      - 32
  - Identifier:
      - 85
      - 115
      - 101
      - 83
      - 116
      - 114
      - 101
      - 116
      - 99
      - 104
      - 66
      - 108
      - 116
  - Whitespace:
      - 32
  - AsKeyword:
      - 65
      - 115
  - Whitespace:
      - 32
  - BooleanKeyword:
      - 66
      - 111
      - 111
      - 108
      - 101
      - 97
      - 110
  - Newline:
      - 13
      - 10
  - Newline:
      - 13
      - 10
